File: code/apps/web/src/lib/orders/mapping.ts
/**
 * Orders protobuf messages to the composer's models.
 */
import { timestampDate } from '@bufbuild/protobuf/wkt';
import type { Timestamp } from '@bufbuild/protobuf/wkt';
import {
	OrderStatus as WireOrderStatus,
	OrderType as WireOrderType,
	Priority as WirePriority,
	type Order as WireOrder
} from '$gen/healthcare/orders/v1/orders_pb.js';
import type { CriticalResult as WireCriticalResult } from '$gen/healthcare/clinical/v1/clinical_pb.js';
import {
	describeOrderStatus,
	type DuplicateCandidate,
	type InboxItem,
	type OrderPriority,
	type OrderStatus,
	type OrderType
} from './composer.js';
import { toPresentedObservation } from '../chart/mapping.js';

function toDate(timestamp: Timestamp | undefined): Date | null {
	return timestamp ? timestampDate(timestamp) : null;
}

const types: Record<WireOrderType, OrderType> = {
	[WireOrderType.UNSPECIFIED]: 'unspecified',
	[WireOrderType.LABORATORY]: 'laboratory',
	[WireOrderType.IMAGING]: 'imaging',
	[WireOrderType.MEDICATION]: 'medication',
	[WireOrderType.PROCEDURE]: 'procedure',
	[WireOrderType.DIET]: 'diet',
	[WireOrderType.NURSING]: 'nursing',
	[WireOrderType.BLOOD_PRODUCT]: 'blood_product',
	[WireOrderType.REFERRAL]: 'referral',
	[WireOrderType.ALLIED_HEALTH]: 'allied_health'
};

const statuses: Record<WireOrderStatus, OrderStatus> = {
	[WireOrderStatus.UNSPECIFIED]: 'unspecified',
	[WireOrderStatus.DRAFT]: 'draft',
	[WireOrderStatus.REQUESTED]: 'requested',
	[WireOrderStatus.ACCEPTED]: 'accepted',
	[WireOrderStatus.SCHEDULED]: 'scheduled',
	[WireOrderStatus.IN_PROGRESS]: 'in_progress',
	[WireOrderStatus.COMPLETED]: 'completed',
	[WireOrderStatus.CANCELLED]: 'cancelled',
	[WireOrderStatus.ENTERED_IN_ERROR]: 'entered_in_error'
};

const priorities: Record<WirePriority, OrderPriority> = {
	[WirePriority.UNSPECIFIED]: 'unspecified',
	[WirePriority.ROUTINE]: 'routine',
	[WirePriority.URGENT]: 'urgent',
	[WirePriority.STAT]: 'stat',
	[WirePriority.TIMING_CRITICAL]: 'timing_critical'
};

/** The wire enum for an order type, for building a request. */
export const wireOrderTypes: Record<OrderType, WireOrderType> = {
	unspecified: WireOrderType.UNSPECIFIED,
	laboratory: WireOrderType.LABORATORY,
	imaging: WireOrderType.IMAGING,
	medication: WireOrderType.MEDICATION,
	procedure: WireOrderType.PROCEDURE,
	diet: WireOrderType.DIET,
	nursing: WireOrderType.NURSING,
	blood_product: WireOrderType.BLOOD_PRODUCT,
	referral: WireOrderType.REFERRAL,
	allied_health: WireOrderType.ALLIED_HEALTH
};

/** The wire enum for a priority, for building a request. */
export const wirePriorities: Record<OrderPriority, WirePriority> = {
	unspecified: WirePriority.UNSPECIFIED,
	routine: WirePriority.ROUTINE,
	urgent: WirePriority.URGENT,
	stat: WirePriority.STAT,
	timing_critical: WirePriority.TIMING_CRITICAL
};

/** What an order row shows. */
export interface PresentedOrder {
	readonly orderId: string;
	readonly number: string;
	readonly display: string;
	readonly type: OrderType;
	readonly status: OrderStatus;
	readonly statusLabel: string;
	readonly priority: OrderPriority;
	readonly indication: string;
	readonly placedAt: Date;
	readonly requesterId: string;
	/** True while the order can still be acted on downstream. */
	readonly live: boolean;
	/** Set when a duplicate override was recorded against this order. */
	readonly overrideReason: string;
	readonly version: bigint;
}

/** Statuses where the order is still somebody's work. */
const liveStatuses = new Set<OrderStatus>([
	'draft',
	'requested',
	'accepted',
	'scheduled',
	'in_progress'
]);

/** Adapts a wire Order. */
export function toPresentedOrder(order: WireOrder): PresentedOrder {
	const status = statuses[order.status] ?? 'unspecified';
	return {
		orderId: order.orderId,
		// The human-facing number, which is what a laboratory quotes on the
		// phone. Falls back to the id rather than showing nothing.
		number: order.number || order.orderId.slice(0, 8),
		display: order.code?.display || order.code?.code || 'Order',
		type: types[order.type] ?? 'unspecified',
		status,
		statusLabel: describeOrderStatus(status),
		priority: priorities[order.priority] ?? 'unspecified',
		indication: order.indication,
		placedAt: toDate(order.createdAt) ?? new Date(0),
		requesterId: order.requesterId,
		live: liveStatuses.has(status),
		overrideReason: order.duplicateOverride?.reason ?? '',
		version: order.version
	};
}

/** Adapts a wire Order into a duplicate candidate row. */
export function toDuplicateCandidate(order: WireOrder): DuplicateCandidate {
	const presented = toPresentedOrder(order);
	return {
		orderId: presented.orderId,
		number: presented.number,
		display: presented.display,
		status: presented.status,
		statusLabel: presented.statusLabel,
		placedAt: presented.placedAt,
		requesterId: presented.requesterId
	};
}

/** Adapts a wire CriticalResult for the inbox. */
export function toInboxItem(result: WireCriticalResult, now: Date): InboxItem | null {
	const observation = result.observation;
	if (!observation) {
		// Nothing to render or act on. Dropped rather than shown as a blank
		// row, which reads as a result whose details failed to load.
		return null;
	}
	const presented = toPresentedObservation(observation);
	return {
		observationId: presented.observationId,
		patientId: observation.patientId,
		display: presented.display,
		value: presented.value,
		interpretationLabel: presented.interpretationLabel,
		effectiveAt: presented.effectiveAt,
		dueEscalations: result.dueEscalations,
		waitingMinutes: Math.max(
			0,
			Math.floor((now.getTime() - presented.effectiveAt.getTime()) / 60_000)
		)
	};
}

File: code/apps/mobile/lib/src/orders/mapping.dart
/// Wire types to the order composer's models (UX-W1-04).
///
/// Same unknown-enum discipline as the other two workspaces; the reasoning is
/// written out in `reception/mapping.dart`.
///
/// The fallback that matters here is the order type. An unreadable type must
/// not become `laboratory`, because the type is what routes the order to a
/// performing service — a misrouted order is work that either never happens or
/// happens in the wrong department. `unrecognised` is refused by
/// `validateOrder`, so the composer will not place one.
library;

import 'package:protobuf/protobuf.dart' as pb;

import '../gen/healthcare/orders/v1/orders.pb.dart' as wire;
import 'composer.dart';

/// Field tags carrying the enums this module reads. Pinned by test.
const int orderTypeField = 3; // orders.v1.Order.type
const int orderPriorityField = 14; // orders.v1.Order.priority
const int orderStatusField = 17; // orders.v1.Order.status

/// True when [message] carried a value for [tag] this build could not read.
bool sentUnknownValueFor(pb.GeneratedMessage message, int tag) =>
    message.unknownFields.hasField(tag);

OrderType orderTypeOf(wire.Order order) {
  if (sentUnknownValueFor(order, orderTypeField)) {
    return OrderType.unrecognised;
  }
  return switch (order.type) {
    wire.OrderType.ORDER_TYPE_LABORATORY => OrderType.laboratory,
    wire.OrderType.ORDER_TYPE_IMAGING => OrderType.imaging,
    wire.OrderType.ORDER_TYPE_MEDICATION => OrderType.medication,
    wire.OrderType.ORDER_TYPE_PROCEDURE => OrderType.procedure,
    wire.OrderType.ORDER_TYPE_DIET => OrderType.diet,
    wire.OrderType.ORDER_TYPE_NURSING => OrderType.nursing,
    wire.OrderType.ORDER_TYPE_BLOOD_PRODUCT => OrderType.bloodProduct,
    wire.OrderType.ORDER_TYPE_REFERRAL => OrderType.referral,
    wire.OrderType.ORDER_TYPE_ALLIED_HEALTH => OrderType.alliedHealth,
    wire.OrderType.ORDER_TYPE_UNSPECIFIED => OrderType.unspecified,
    _ => OrderType.unrecognised,
  };
}

OrderStatus orderStatusOf(wire.Order order) {
  if (sentUnknownValueFor(order, orderStatusField)) {
    return OrderStatus.unrecognised;
  }
  return switch (order.status) {
    wire.OrderStatus.ORDER_STATUS_DRAFT => OrderStatus.draft,
    wire.OrderStatus.ORDER_STATUS_REQUESTED => OrderStatus.requested,
    wire.OrderStatus.ORDER_STATUS_ACCEPTED => OrderStatus.accepted,
    wire.OrderStatus.ORDER_STATUS_SCHEDULED => OrderStatus.scheduled,
    wire.OrderStatus.ORDER_STATUS_IN_PROGRESS => OrderStatus.inProgress,
    wire.OrderStatus.ORDER_STATUS_COMPLETED => OrderStatus.completed,
    wire.OrderStatus.ORDER_STATUS_CANCELLED => OrderStatus.cancelled,
    wire.OrderStatus.ORDER_STATUS_ENTERED_IN_ERROR => OrderStatus.enteredInError,
    wire.OrderStatus.ORDER_STATUS_UNSPECIFIED => OrderStatus.unspecified,
    _ => OrderStatus.unrecognised,
  };
}

OrderPriority orderPriorityOf(wire.Order order) {
  if (sentUnknownValueFor(order, orderPriorityField)) {
    return OrderPriority.unrecognised;
  }
  return switch (order.priority) {
    wire.Priority.PRIORITY_STAT => OrderPriority.stat,
    wire.Priority.PRIORITY_URGENT => OrderPriority.urgent,
    wire.Priority.PRIORITY_TIMING_CRITICAL => OrderPriority.timingCritical,
    wire.Priority.PRIORITY_ROUTINE => OrderPriority.routine,
    wire.Priority.PRIORITY_UNSPECIFIED => OrderPriority.unspecified,
    _ => OrderPriority.unrecognised,
  };
}

/// Wire order type for a model one, for placing.
///
/// Returns null for anything the composer should not be placing under, which
/// `validateOrder` has already refused — this is the second reading of the
/// same rule, at the point where it would otherwise reach the wire.
wire.OrderType? wireOrderTypeOf(OrderType type) => switch (type) {
      OrderType.laboratory => wire.OrderType.ORDER_TYPE_LABORATORY,
      OrderType.imaging => wire.OrderType.ORDER_TYPE_IMAGING,
      OrderType.medication => wire.OrderType.ORDER_TYPE_MEDICATION,
      OrderType.procedure => wire.OrderType.ORDER_TYPE_PROCEDURE,
      OrderType.diet => wire.OrderType.ORDER_TYPE_DIET,
      OrderType.nursing => wire.OrderType.ORDER_TYPE_NURSING,
      OrderType.bloodProduct => wire.OrderType.ORDER_TYPE_BLOOD_PRODUCT,
      OrderType.referral => wire.OrderType.ORDER_TYPE_REFERRAL,
      OrderType.alliedHealth => wire.OrderType.ORDER_TYPE_ALLIED_HEALTH,
      OrderType.unspecified || OrderType.unrecognised => null,
    };

/// Wire priority for a model one. Unspecified where there is nothing to say.
wire.Priority wirePriorityOf(OrderPriority priority) => switch (priority) {
      OrderPriority.stat => wire.Priority.PRIORITY_STAT,
      OrderPriority.urgent => wire.Priority.PRIORITY_URGENT,
      OrderPriority.timingCritical => wire.Priority.PRIORITY_TIMING_CRITICAL,
      OrderPriority.routine => wire.Priority.PRIORITY_ROUTINE,
      OrderPriority.unspecified ||
      OrderPriority.unrecognised =>
        wire.Priority.PRIORITY_UNSPECIFIED,
    };

/// One existing order, as a duplicate candidate.
DuplicateCandidate candidateOf(wire.Order order) => DuplicateCandidate(
      orderId: order.orderId,
      number: order.number,
      display: order.code.display.isNotEmpty
          ? order.code.display
          : order.code.code,
      status: orderStatusOf(order),
      // created_at, because that is when the order was placed — the contract
      // has no separate placed_at, and the status history's first entry would
      // be the same instant by a longer route.
      placedAt: order.createdAt.toDateTime().toUtc(),
      requesterId: order.requesterId,
    );

/// The duplicate gate for a warning the server returned.
DuplicateGate gateForWarning(
  wire.DuplicateWarning warning, {
  List<String> acknowledged = const [],
  String reason = '',
}) =>
    duplicateGate(
      candidates: [for (final o in warning.existing) candidateOf(o)],
      overridable: warning.overridable,
      windowSeconds: warning.windowSeconds.toInt(),
      acknowledged: acknowledged,
      reason: reason,
    );
